external-crates/move/move-compiler/src/sui_mode/typing.rs

```diff
@@ -0,0 +1,434 @@
+}
+
+fn is_entry_primitive_ty(param_ty: &Type) -> bool {
+    use BuiltinTypeName_ as B;
+    use TypeName_ as N;
+
+    match &param_ty.value {
+        // A bit of a hack since no primitive has key
+        Type_::Param(tp) => !tp.abilities.has_ability_(Ability_::Key),
+        // nonsensical, but no error needed
+        Type_::Apply(_, sp!(_, N::Multiple(_)), ts) => ts.iter().all(is_entry_primitive_ty),
+        // Simple recursive cases
+        Type_::Ref(_, t) => is_entry_primitive_ty(t),
+        Type_::Apply(_, sp!(_, N::Builtin(sp!(_, B::Vector))), targs) => {
+            debug_assert!(targs.len() == 1);
+            is_entry_primitive_ty(&targs[0])
+        }
+
+        // custom "primitives"
+        Type_::Apply(_, sp!(_, n), targs)
+            if n.is(STD_ADDR_NAME, ASCII_MODULE_NAME, ASCII_TYPE_NAME)
+                || n.is(STD_ADDR_NAME, UTF_MODULE_NAME, UTF_TYPE_NAME)
+                || n.is(SUI_ADDR_NAME, OBJECT_MODULE_NAME, ID_TYPE_NAME) =>
+        {
+            debug_assert!(targs.is_empty());
+            true
+        }
+        Type_::Apply(_, sp!(_, n), targs)
+            if n.is(STD_ADDR_NAME, OPTION_MODULE_NAME, OPTION_TYPE_NAME) =>
+        {
+            debug_assert!(targs.len() == 1);
+            is_entry_primitive_ty(&targs[0])
+        }
+
+        // primitives
+        Type_::Apply(_, sp!(_, N::Builtin(_)), targs) => {
+            debug_assert!(targs.is_empty());
+            true
+        }
+
+        // Non primitive
+        Type_::Apply(_, sp!(_, N::ModuleType(_, _)), _) => false,
+        Type_::Unit => false,
+
+        // Error case nothing to do
+        Type_::UnresolvedError | Type_::Anything | Type_::Var(_) => true,
+    }
+}
+
+fn is_entry_object_ty(param_ty: &Type) -> bool {
+    use BuiltinTypeName_ as B;
+    use TypeName_ as N;
+    match &param_ty.value {
+        Type_::Ref(_, t) => is_entry_object_ty_inner(t),
+        Type_::Apply(_, sp!(_, N::Builtin(sp!(_, B::Vector))), targs) => {
+            debug_assert!(targs.len() == 1);
+            is_entry_object_ty_inner(&targs[0])
+        }
+        _ => is_entry_object_ty_inner(param_ty),
+    }
+}
+
+fn is_entry_object_ty_inner(param_ty: &Type) -> bool {
+    use TypeName_ as N;
+    match &param_ty.value {
+        Type_::Param(tp) => tp.abilities.has_ability_(Ability_::Key),
+        // nonsensical, but no error needed
+        Type_::Apply(_, sp!(_, N::Multiple(_)), ts) => ts.iter().all(is_entry_object_ty_inner),
+        // Simple recursive cases, shouldn't be hit but no need to error
+        Type_::Ref(_, t) => is_entry_object_ty_inner(t),
+
+        // Objects
+        Type_::Apply(Some(abilities), _, _) => abilities.has_ability_(Ability_::Key),
+
+        // Error case nothing to do
+        Type_::UnresolvedError | Type_::Anything | Type_::Var(_) | Type_::Unit => true,
+        // Unreachable cases
+        Type_::Apply(None, _, _) => unreachable!("ICE abilities should have been expanded"),
+    }
+}
+
+fn entry_return(
```

Comment:
Can we document or link to the documentation for what we expect from a valid entry return type? (TIL we can return `drop`-able types from an entry functions).

Reply:
I can do it as a separate PR, sure!